cl/recap/mergers.py

```diff
@@ -607,28 +604,18 @@ def get_or_make_docket_entry(
      - None is returned when things fail.
     """
     if docket_entry["document_number"]:
-        with transaction.atomic():
-            # In order to create a lock, we need to retrieve the docket using
-            # select_for_update
-            docket = Docket.objects.select_for_update().get(pk=d.pk)
```

Comment:
I'm concerned that removing the database lock implemented through `select_for_update` might lead to duplicate docket entries. This lock was initially added to address an [issue](https://github.com/freelawproject/courtlistener/issues/2144) with duplicated docket entries caused by a race condition, solved in https://github.com/freelawproject/courtlistener/pull/2299

I've managed to reproduce the issue using this new approach. Now that there is no database lock on the docket row where docket entries are being added, it's possible for two different processes to add the same entries.
![Screenshot 2023-09-08 at 14 22 27](https://github.com/freelawproject/courtlistener/assets/486004/4f8b62e1-9a2a-4553-a684-50d937c69cc7)

Perhaps moving the entire transaction, including the lock, into a synchronous method could resolve the issue? We could then wrap this synchronous method with `sync_to_async(sync_db_transaction_and_lock)(...)`

Reply:
> I've managed to reproduce the issue using this new approach. Now that there is no database lock on the docket row where docket entries are being added, it's possible for two different processes to add the same entries.

Hmm, how would I reproduce this exactly?

> Perhaps moving the entire transaction, including the lock, into a synchronous method could resolve the issue?

I'm trying to avoid a lock for performance reasons. I thought `aget_or_create` can be used to avoid the need for a lock here.

Reply:
Sure these are the steps to reproduce the race condition:

- You'd need to submit the same docket with multiple concurrently requests, for example here 10 concurrent curl requests:

```
   seq 1 10 | xargs -n1 -P10  curl --location --request POST 'http://localhost:8000/api/rest/v3/recap/' \
--header 'Authorization: Token 0593fd1acbc56b250362653d17707b09bbc04210' \
--form 'upload_type="1"' \
--form 'filepath_local=@"nysd_1_docket_simplify.html"' \
--form 'court="nysd"' \
--form 'pacer_case_id="468494"' \
--form 'docket_number="1:17-cv-00916"'
```
It's not sure that you get the race condition at the first attempt, you'd need to try multiple times.
In order to increase the opportunity to observe the race condition I do the following:

- Go to the submitted docket and remove all of its docket entries. This way, when you resubmit the docket, only the new docket entries will be created.

- I recommend using a docket with a single docket entry, I'm attaching the example I use.
[nysd_1_docket_simplify.txt](https://github.com/freelawproject/courtlistener/files/12564101/nysd_1_docket_simplify.txt)

Reply:
I think I found the root cause of the race issue with [`get_or_create`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#get-or-create).

> This method is atomic assuming that the database enforces uniqueness of the keyword arguments (see [unique](https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.Field.unique) or [unique_together](https://docs.djangoproject.com/en/4.2/ref/models/options/#django.db.models.Options.unique_together)). If the fields used in the keyword arguments do not have a uniqueness constraint, concurrent calls to this method may result in multiple rows with the same parameters being inserted.

We're missing a `unique_together` constraint, I've gone ahead and added that.

Reply:
Boy, that feels like a big solution. I'm not sure it's true either, for our DB, that these are always unique together. Why not keep the select_for_update?

Reply:
Oh yes, the unique constraint addresses the race condition in this situation. However, when we were [discussing](https://github.com/freelawproject/courtlistener/issues/2144#issuecomment-1195847274) this issue, we decided not to add the constraint. To implement it, we would first need to remove all the duplicate `docket-entry_number` entries in the database. Upon inspection, we found that we already have docket entries with the same `entry_number` that are not actually duplicates. At that time, we concluded that it would be too risky to remove all of those entries. As a result, we opted to use database locking instead.

Reply:
> I'm not sure it's true either, for our DB, that these are always unique together.

The code at least seems to want these to be unique from the looks of it and doesn't handle non-uniques properly.

> Why not keep the select_for_update?

From my understanding this is a pretty big performance footgun.

> Upon inspection, we found that we already have docket entries with the same `entry_number` that are not actually duplicates.

Hmm, do we have a full list of these somewhere?

Reply:
No, no list for the moment. I agree it's not the best for performance, but I don't think it's hurting us too much at this scale.

Reply:
> I don't think it's hurting us too much at this scale.

Well this function is called from a view handler directly so it needs to be reasonably fast for large dockets I think.

I've gone ahead and switched back to using `select_for_update` and also added some cleanup handling for when duplicates are encountered.

Reply:
Great. Back to you, then, @albertisfu!